lib/shoryuken/polling.rb

```diff
@@ -102,5 +123,94 @@ def queue_weight(queues, queue)
         queues.count { |q| q == queue }
       end
     end
+
+    class StrictPriority < BaseStrategy
+
+      def initialize(queues)
+        # Mapping of queues to priority values
+        @queue_priorities = queues
+          .each_with_object(Hash.new(0)) { |queue, h| h[queue] += 1 }
+
+        # Priority ordering of the queues
+        @queue_order = @queue_priorities
+          .to_a
+          .sort_by { |queue, priority| -priority }
+          .map(&:first)
+
+        # Pause status of the queues
+        @queue_status = @queue_order
+          .each_with_object({}) { |queue, h| h[queue] = [true, nil] }
+
+        # Most recently used queue
+        @current_queue = nil
+      end
+
+      def next_queue
+        unpause_queues
+        @current_queue = next_active_queue
+        return nil if @current_queue.nil?
+        QueueConfiguration.new(@current_queue, {})
+      end
+
+      def messages_found(queue, messages_found)
+        if messages_found == 0
+          # If no messages are found, we pause a given queue
+          pause(queue) 
+        else
+          # Reset the current queue when messages found to cause priorities to re-run
+          @current_queue = nil
+        end
+      end
+
+      def active_queues
+        @queue_status
+          .select { |_, status| status.first }
+          .map { |queue, _| [queue, @queue_priorities[queue]] }
+      end
+
+      private
+
+      def next_active_queue
+        return nil unless @queue_order.length > 0
+
+        start = @current_queue.nil? ? 0 : @queue_order.index(@current_queue) + 1
+        i = 0
+
+        # Loop through the queue order from the current queue until we find a
+        # queue that is next in line and is not paused
+        while true
```

Comment:
It's not drop-in but we have similar code in our custom stuff that works like this:

```ruby
queues = @queue_order.dup

while true
  queue = queues.first
  # process queue
  queues.rotate!
end
```

So rather than maintaining `i` or `@current_queue` you `dup` the ordered queues at some point and then `Array#rotate!` them as needed to move to the next queue.